README.md
# flutter_sound_record

Audio recorder from microphone to a given file path forked from and based on <https://github.com/llfbandit/record>.

No external dependencies, MediaRecorder is used for Android and AVAudioRecorder for iOS.

## Options

- bit rate (be careful with this one on iOs)
- sampling rate
- encoder

## Platforms

### Android

```xml
<uses-permission android:name="android.permission.RECORD_AUDIO" />
<!-- Optional, you'll have to check this permission by yourself. -->
<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
```

min SDK: 16 (29 if you use OPUS)

### iOs

```xml
<key>NSMicrophoneUsageDescription</key>
<string>We need to access to the microphone to record audio file</string>
```

min SDK: 8.0 (11 if you use OPUS)

## Supported encoders

```dart

enum AudioEncoder {
  /// Will output to MPEG_4 format container
  AAC,

  /// Will output to MPEG_4 format container
  AAC_LD,

  /// Will output to MPEG_4 format container
  AAC_HE,

  /// sampling rate should be set to 8kHz
  /// Will output to 3GP format container on Android
  AMR_NB,

  /// sampling rate should be set to 16kHz
  /// Will output to 3GP format container on Android
  AMR_WB,

  /// Will output to MPEG_4 format container
  /// /!\ SDK 29 on Android /!\
  /// /!\ SDK 11 on iOs /!\
  OPUS,
}
```

### Android

<https://developer.android.com/reference/android/media/MediaRecorder.AudioEncoder>

### iOs
<https://developer.apple.com/documentation/coreaudiotypes/coreaudiotype_constants/1572096-audio_data_format_identifiers>

## Usage

```dart
// Import package
import 'package:record/record.dart';

// Check and request permission
bool result = await Record.hasPermission();

// Start recording
await Record.start(
  path: 'aFullPath/myFile.m4a', // required
  encoder: AudioEncoder.AAC, // by default
  bitRate: 128000, // by default
  sampleRate: 44100, // by default
);

// Stop recording
await Record.stop();

// Get the state of the recorder
bool isRecording = await Record.isRecording();

// There's nothing to dispose, this done internally each time you call stop method.
// The plugin is aware of activity lifecycle.
// So exiting, your app or activity will stop the recording (but won't delete the
// output file).
```

## Warnings

Be sure to check supported values from the given links above.

## Known issues

None.


CHANGELOG.md
# Changelog

## 3.3.2

* fix: Depend on new flutter_sound_record_web version

## 3.3.1

* fix: Add support for AGP 8.0

## 3.2.1

* core: Update dependencies.

## 3.2.0

* core: convert to kotlin

## 3.1.2

* fix: Use corrected platform interface.

## 3.1.1

* fix: Eleminate some deprecated functions and classes.
* core: Rebrand.

## 3.1.0

* fix: get build working on Android SDK 31
* core: change project structure
* core: add linter rules, VS Code settings and run configuration

## 3.0.3

* core: Moved from jcenter to mavenCentral (Android).

## 3.0.2

* fix: web `pubspec.yaml` file.

## 3.0.1

* fix: Android request code is now frozen.
* core: Update dependencies.

## 3.0.0

* feat: Add web support.
* feat: Add `getAmplitude()` to get current and max dBFS.
* core: static methods are no more! (but api is pretty much the same)
* core: path param is now optional in `start` method to align behaviour with web platform.
If null is given, a temp file is generated for you.
* core: `stop` method now returns the output path.

## 2.1.1

* fix: Android warning "mediarecorder went away with unhandled events".

## 2.1.0+1

* Relaxing dart/flutter constraints for pub.dev analysis.

## 2.1.0

* Add pause/resume/isPaused features.
* Updated example to include pause.

## 2.0.0

* Add null safety support (based on v1.0.3).
* Updated example. Moved from 'audioplayers' to 'just_audio'.

## 1.0.3

* Allow recording from bluetooth device on iOS.

## 1.0.2

* Fix (for good) potential exception when closing beofre recorder is actually started (Android).

## 1.0.1

* Fix potential exception when closing beofre recorder is actually started (Android).
* Fix potential NPE on permission result (Android).
* Updade dependencies.

## 1.0.0

* Flutter 1.12.0 is now minimum version (Flutter plugin API v2).
* Finalize Record API.
* Request permission on Android too.
* Automatically stop recording when leaving activity / app.
* Add all codecs with cross platform compatibility.

## 0.2.1

* Fix broken build on Android.

## 0.2.0

* Fix broken build on Android.

## 0.1.0

* Initial release.
* Implementations for Android and iOs.
* Example.


ios/Classes/SwiftFlutterRecordPlugin.swift
import Flutter
import UIKit
import AVFoundation

public class SwiftFlutterRecordPlugin: NSObject, FlutterPlugin, AVAudioRecorderDelegate {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "com.josephcrowell.flutter_record", binaryMessenger: registrar.messenger())
    let instance = SwiftFlutterRecordPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
    registrar.addApplicationDelegate(instance)
  }

  var isRecording = false
  var isPaused = false
  var hasPermission = false
  var audioRecorder: AVAudioRecorder?
  var path: String?
  var maxAmplitude:Float = -160.0;

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
      case "start":
        let args = call.arguments as! [String : Any]
        path = args["path"] as? String

        if path == nil {
          let directory = NSTemporaryDirectory()
          let fileName = UUID().uuidString + ".m4a"

            path = NSURL.fileURL(withPathComponents: [directory, fileName])?.absoluteString
        }

        start(
          path: path!,
          encoder: args["encoder"] as? Int ?? 0,
          bitRate: args["bitRate"] as? Int ?? 128000,
          samplingRate: args["samplingRate"] as? Float ?? 44100.0,
          result: result);
        break
      case "stop":
        stop(result)
        break
      case "pause":
        pause(result)
      case "resume":
        resume(result)
      case "isPaused":
        result(isPaused)
      case "isRecording":
        result(isRecording)
        break
      case "hasPermission":
        hasPermission(result)
        break
      case "getAmplitude":
        getAmplitude(result)
        break
      case "dispose":
        dispose(result)
        break
      default:
        result(FlutterMethodNotImplemented)
        break
    }
  }

  public func applicationWillTerminate(_ application: UIApplication) {
    stopRecording()
  }

  public func applicationDidEnterBackground(_ application: UIApplication) {
    stopRecording()
  }

  fileprivate func hasPermission(_ result: @escaping FlutterResult) {
    switch AVAudioSession.sharedInstance().recordPermission {
      case AVAudioSession.RecordPermission.granted:
        hasPermission = true
        break
      case AVAudioSession.RecordPermission.denied:
        hasPermission = false
        break
      case AVAudioSession.RecordPermission.undetermined:
        AVAudioSession.sharedInstance().requestRecordPermission() { [unowned self] allowed in
          DispatchQueue.main.async {
            self.hasPermission = allowed
          }
        }
        break
      default:
        break
    }

    result(hasPermission)
  }

  fileprivate func start(path: String, encoder: Int, bitRate: Int, samplingRate: Float, result: @escaping FlutterResult) {
    stopRecording()

    let settings = [
      AVFormatIDKey: getEncoder(encoder),
      AVEncoderBitRateKey: bitRate,
      AVSampleRateKey: samplingRate,
      AVNumberOfChannelsKey: 2,
      AVEncoderAudioQualityKey: AVAudioQuality.high.rawValue
    ] as [String : Any]

    let options: AVAudioSession.CategoryOptions = [.defaultToSpeaker, .allowBluetooth]

    do {
      try AVAudioSession.sharedInstance().setCategory(AVAudioSession.Category.playAndRecord, options: options)
      try AVAudioSession.sharedInstance().setActive(true)

      let url = URL(string: path) ?? URL(fileURLWithPath: path)
      audioRecorder = try AVAudioRecorder(url: url, settings: settings)
      audioRecorder!.delegate = self
      audioRecorder!.isMeteringEnabled = true
      audioRecorder!.record()

      isRecording = true
      isPaused = false
      result(nil)
    } catch {
      result(FlutterError(code: "", message: "Failed to start recording", details: nil))
    }
  }

  fileprivate func stop(_ result: @escaping FlutterResult) {
    stopRecording()
    result(path)
  }

  fileprivate func pause(_ result: @escaping FlutterResult) {
    audioRecorder?.pause()
    isPaused = true
    result(nil)
  }

  fileprivate func resume(_ result: @escaping FlutterResult) {
    if isPaused {
      audioRecorder?.record()
      isPaused = false
    }

    result(nil)
  }

  fileprivate func getAmplitude(_ result: @escaping FlutterResult) {
    var amp = [String : Float]()

    amp["current"] = -160.0

    if isRecording {
      audioRecorder?.updateMeters()

      let current = audioRecorder?.averagePower(forChannel: 0)

      if (current! > maxAmplitude) {
        maxAmplitude = current!;
      }

      amp["current"] = current
      amp["max"] = maxAmplitude
    }

    result(amp)
  }

  fileprivate func stopRecording() {
    audioRecorder?.stop()
    audioRecorder = nil
    isRecording = false
    isPaused = false
    maxAmplitude = -160.0
  }

  fileprivate func dispose(_ result: @escaping FlutterResult) {
    stopRecording()
    result(path)
  }

  // https://developer.apple.com/documentation/coreaudiotypes/coreaudiotype_constants/1572096-audio_data_format_identifiers
  fileprivate func getEncoder(_ encoder: Int) -> Int {
    switch(encoder) {
    case 1:
      return Int(kAudioFormatMPEG4AAC_ELD)
    case 2:
      return Int(kAudioFormatMPEG4AAC_HE)
    case 3:
      return Int(kAudioFormatAMR)
    case 4:
      return Int(kAudioFormatAMR_WB)
    case 5:
      return Int(kAudioFormatOpus)
    default:
      return Int(kAudioFormatMPEG4AAC)
    }
  }
}
